### PR TITLE
Parse synonyms with the same analysis chain

### LINKED ISSUE #7199: Parse synonyms with the same analysis chain
Synonyms are currently tokenized with the whitespace tokenizer.  Really they should be tokenized with whatever tokenizer and token filters appear before it in the chain (ie the same analysis chain that is used on the text).

### PR DESCRIPTION
**Pull request by johtani**
Change behavior synonym filter factory.
Now, synonyms are tokenized with the whitespace tokenizer or a tokenizer specified "tokenizer" parameter.
This PR is to tokenize synonyms with whatever tokenizer and token filters appear before it in the chain.
- change parsing synonyms from constructor to building custom analyzer
- clone SynonymTokenFilterFactory with the analysis chain in CustomAnalyzerProvider

Close #7199

### CODE REVIEW

**Comment by markharwood** [comment]
Just downloaded to take a look @johtani and this has a compile error against master due to changes in Lucene's Analyzer in LUCENE-5388

**Comment by johtani** [comment]
Oh... I will rebase and fix it.

**Comment by johtani** [comment]
@markharwood Rebase and fix LUCENE-5388. please review again.

**Comment by clintongormley** [comment]
@markharwood please could you review

**Comment by markharwood** [comment]
Should the `ignore_case` setting be removed/deprecated too? I can see it was a requirement when we relied on WhitespaceTokenizer but if we shift to using the configured token stream it just becomes extra baggage?

**Comment by johtani** [comment]
Thanks for your comment, @markharwood .
Good point.
I think we should `ignore_case` and `tokenizer` setting be deprecated.
I will add the description to document.

**Comment by markharwood** [comment]
In testing I came across this - see https://gist.github.com/markharwood/a69d055e8027f2a8a865

Note the offsets and positions of the output synonyms - things start to go wrong when we inject overlapping and parallel streams of tokens as synonyms. 
There is no sensible way of representing multiple output synonyms of variable numbers of words and retaining a sensible notion of offset and position.
Each output synonym effectively creates an alternative branch which has a token numbering system that is incompatible with other branches.
Any sense of order (e.g. X follows Y) can no longer be reasoned about when we fork these alternative streams in a system designed to record only one sequence - it's a bit like trying to keep track of the plot of Back To The Future movies.
@clintongormley  Maybe we should assume that output synonyms always have to be single tokens?


Also, note in my Gist the change in behaviour of synonym entry "Cosmos" now being output as "cosmos". I guess a more common scenario would be an output token like `mp3_player` would typically now be output as `mp3` and `player`. This is potentially a breaking change in behaviour - do we 
a) document this breaking change in our migration guide or
b) consider what options are needed to maintain backwards compatibility?

I haven't seen enough synonym files in the wild to comment but I think these 2 points may be connected - people may reasonably expect something like output synonym `mp3_player` to be untokenized because  it avoids the danger of false matches on `player` and doesn't introduce alternative streams of ordering like the confusion time-travel adds to any movie plot.

**Comment by markharwood** [comment]
Further on this - the wise old owl that is @mikemccand had this:

    "you should ensure all of your injected synonyms are single tokens! " [1]

Maybe it is a mistake for us to introduce this change?

[1] http://blog.mikemccandless.com/2012/04/lucenes-tokenstreams-are-actually.html

**Comment by rmuir** [comment]
I don't think its a mistake. Instead of the user worrying about two things:
1) putting synonymfilter in the correct place in the chain
2) configuring synonymfilter's own analysis process to be correct

Users just need to put synonymfilter in the correct place in their analysis chain. We can't fix the first here. But it removes the 2nd and makes it easier to use.

**Comment by markharwood** [comment]
There's 3 token streams to consider:
1) The doc's terms
2) The synonym rule input condition terms
3) The synonym rule output terms

Agreed, it is convenient if 1) and 2) are made to talk the same language.
Disagree about the appropriateness of tokenization on 3) for the reasons Mike outlined in his blog.

**Comment by rmuir** [comment]
I dont think the user should have to consider the third either. Again its just a matter of inserting the synonymfilter in the correct place in the analysis chain, so that both input and output processing makes sense.

**Comment by rmuir** [comment]
Just think about it: if the synonyms don't go through the same output processing, then they are useless always (they will not match with expected terms because processing is different).

Mike's blog isn't really relevant here. Either you have synonyms in the analysis chain and they emit terms consistent with the way the analysis chain processes all text, or don't use synonymfilter.

**Comment by markharwood** [comment]
Just think about it: if the synonyms don't go through the same output processing,
    then they are useless 

Not always. In a normalization rule the output term doesn't have to be anything other than a single indexed token that represents a common concept mapped to various forms of end-user expressions e.g.

    i pod, i-pod, mp3 player => SKU_3672

Here, end users will not search for SKU_3672 directly - tokenization in this case would be irrelevant and unhelpful.

The alternative use case is an expansion rule where we have 2 use cases - 
1) Index-time expansion of document terms (anticipating future user searches)
2) Query-time expansion of user search criteria

In the case of 1) Mike's blog is relevant and talks about introducing false positives in searches. I note the offsets in my GIST look screwed too so highlighting accuracy might be a concern [1] we need to fix here.
Given these issues and the general inflexibility of the index-time approach an end user might just as well consider using a normalization rule here?

In the case of 2) we can benefit from increased matching capabilities on previously untreated content but I wonder if these outputs are better expressed as actual query expressions if that is their intention? That brings the possibility of adding NOTs etc to the definition.

[1] Check out the reported offset of the word "device" in both master and this PR.

**Comment by markharwood** [comment]
More detailed acknowledgement of the broken-ness to injecting multi-word synonyms: https://www.elastic.co/guide/en/elasticsearch/guide/current/multi-word-synonyms.html

    "The way to avoid this mess is to use simple contraction to inject a single term that represents
    all synonyms" 

With this in mind I'm not sure why we are choosing to switch to a policy that is likely to increase multi-word outputs and could break configurations where users have followed our advice and created single output tokens like 'MP3_player'. By moving away from whitespace tokenizer these tokens would now be split into multiple by the typical tokenizer used for document content.

**Comment by rmuir** [comment]
> Here, end users will not search for SKU_3672 directly - tokenization in this case would be irrelevant and unhelpful.

this is esoteric usage.

**Comment by markharwood** [comment]
this is esoteric usage.

Setting aside the issue of if the output token is likely to be in the end-user's vocabulary or not - single-token output was the recommended solution in both of the articles I referenced. Mike's given example output token was `dns` and Zach/Clint's was `usa`. My pessimistic assumption was not all concepts we would want to represent would be lucky enough to have a single-token representation that users would stand a chance of typing which is why I opted for something that resembled an internal primary key. 

The choice of symbol doesn't change the reasoning for why single-token output is being recommended here.

**Comment by johtani** [comment]
I think multi-word synonym is related LUCENE-6582 https://issues.apache.org/jira/plugins/servlet/mobile#issue/LUCENE-6582

**Comment by johtani** [comment]
@markharwood @rmuir 
Thanks for your discussion.
I think multi-word synonym is relate LUCENE-6582.

And I think "Cosmos" and "mp3_player" is related `tokenizer` option.
If user needs backwards compatibility, then user set `tokenizer: "whitespace"`.
I should to add this to migration guide, right?

**Comment by markharwood** [comment]
Good to see there's some upstream activity in Lucene to tackle the multi-word concerns. 

+1 for documenting `tokenizer: "whitespace"` in the migration guide.

**Comment by johtani** [comment]
Added to the migration guide

**Comment by johtani** [comment]
Related another Lucene ticket.
https://issues.apache.org/jira/browse/LUCENE-6664

**Comment by clintongormley** [comment]
@johtani Where are we with this PR?

**Comment by clintongormley** [comment]
@johtani can we get this in for 5.0 please?

**Comment by paulintrognon** [comment]
Our team has been struggling with multi-terms synonyms for a long time now, and would love to see this feature as part of ElasticSearch's product.

**Comment by dakrone** [comment]
@johtani ping again about updating/merging this PR, can you take a look?

**Comment by johtani** [comment]
@dakrone @clintongormley Sorry for late reply. I will update/rebase the PR.  
I think we should create another PR related to https://issues.apache.org/jira/browse/LUCENE-6664

**Comment by dakrone** [comment]
@johtani it looks like that feature is still not in, we could probably get this one in and then update it once that does get in to Lucene?

**Comment by johtani** [comment]
@dakrone Ah, that's right. 
I think this PR does not solve multi-terms synonym's problem completely, however, this PR is useful to make it easier to use.
We should create the issue about multi-terms synonyms and comment "related to https://issues.apache.org/jira/browse/LUCENE-6664".

Does it make sense?

**Comment by dakrone** [comment]
@johtani sounds good to me!

**Comment by johtani** [comment]
@dakrone  Rebase this. I think need to review again. 
And I found #10394 , we already have an issue of multi-word synonym.

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: I don't think we should swallow the exceptions here, instead, this could be a try-with-resources block like:

```java
try (BufferedReader br = new BufferedReader(rulesReader)) {
    .. read the string ..
}
```

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: This could be `Strings.hasLength(tokenizerName)`

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: Can you move these class variable definitions up to the top of the class? It's weird to see them after function definitions

File: docs/reference/analysis/tokenfilters/synonym-tokenfilter.asciidoc
Comment (dakrone) [inline docs/reference/analysis/tokenfilters/synonym-tokenfilter.asciidoc]: And -> The

File: docs/reference/migration/migrate_5_0/mapping.asciidoc
Comment (dakrone) [inline docs/reference/migration/migrate_5_0/mapping.asciidoc]: "still leave" -> "are still left"

File: docs/reference/migration/migrate_5_0/mapping.asciidoc
Comment (dakrone) [inline docs/reference/migration/migrate_5_0/mapping.asciidoc]: "same behavior in 1.x" -> "will have the same behavior as 1.x"

**Comment by dakrone** [comment]
I left a few comments, but I'd feel better if someone more familiar with analysis took a look at it, maybe @rmuir ?

**Comment by dakrone** [comment]
Not sure who should take a look at this, maybe @jpountz ?

**Comment by johtani** [comment]
@dakrone Sorry... I missed your comment. Rebased master and fix your comment.
Can @jpountz  take a look if you have time?

**Comment by clintongormley** [comment]
ping @jpountz

**Comment by rjernst** [comment]
Should this PR be rethought given the new SynonymGraphFilter added in https://issues.apache.org/jira/browse/LUCENE-6664?

**Comment by mikemccand** [comment]
I think this PR is still very much relevant, even after `SynonynGraphTokenFilter`: it's really vital that the same analysis components preceding the synonym filter (be it the legacy `SynonymFilter` or the new `SynonymGraphFilter`) in the pipeline are used when parsing the incoming synonym rules.

**Comment by johtani** [comment]
Rebase and support SynonymGraphFilterFactory

**Comment by jpountz** [comment]
Thanks @johtani I'll do reviews quickly.

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: This is not good for backword compatibility. Instead it should do:

```
if (indexSettings.getIndexVersionCreated().before(Version.V_6_0_0)) {
    String tokenizerName = settings.get("tokenizer", "whitespace");
    tokenizerFactory = ...;
} else {
    tokenizerFactory = null;
}
```

**Review by jpountz — CHANGES_REQUESTED** [review]
I like the change in general. I left some comments about the implementation.

**Comment by johtani** [comment]
@jpountz Thx for your comment. Fix your comments. 
I think we should add the backward compatibility for tokenizer param.

**Comment by johtani** [comment]
@elasticmachine test this please

**Comment by johtani** [comment]
Rebase master and move analyze test case to rest test

**Comment by johtani** [comment]
@jpountz Finally, CI pass...
Rebase master, then move some Unit test case to rest-api.
If you have time, please review again.
If you need to squash for reviewing, please let me know.

**Comment by rjernst** [comment]
@jpountz Can you review again?
@johtani There are merge conflicts, can you fix?

We should definitely get this in for 6.0, this PR has been open for 2.5 years :)

**Comment by johtani** [comment]
@rjernst Completely agree :) I will fix the conflict today!

**Comment by johtani** [comment]
pushed rebased version and waiting for CI successful

**Comment by johtani** [comment]
@jpountz Rebased and passed CI. please review again if you have time.

File: core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java]: I think it would be better to return a list rather than having side-effects on a list that is passed as an argument

File: core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java]: let's do `for (AnalyzeRequest.NameOrDefinition charFilter : charFilters)` now that we do not need the index anymore

File: core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: this method does not need the gaps I think? Let's remove them from the signature and use the defaults in the body of this method?

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymGraphTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymGraphTokenFilterFactory.java]: :+1:

**Review by jpountz — CHANGES_REQUESTED** [review]
Thanks @johtani I think it is the right way to implement this change. I left some minor comments. However I think this PR should also deprecate `ignoreCase` since it is not useful anymore now that we figure it out automatically based on the analysis chain?

File: core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java
Comment (johtani) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: I'm not sure what you say. The value from https://github.com/elastic/elasticsearch/pull/8049/files/a297d89134c18fad0daf69b2c48356d4c09a6764#diff-1ea188dabb8bc41bc2812d1564604899R84 is not the default value. Should not use the value from analyze_settings always?

Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: What I mean is that gaps only matter when there are multiple field values. In that case, the analyzer will always analyze a single value at a time, so the value of the gaps do not matter. We can pass whatever value we want and clean up the signature of this method.

Comment (johtani) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: Oh... I see! thx

**Comment by johtani** [comment]
@jpountz Thanks for your comments. Fixed deprecation log and small comments.

**Comment by johtani** [comment]
@elasticmachine test this please

**Comment by johtani** [comment]
Rebase master because some qa test failed...

File: core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: we do not need the `positionIncrementGap` either?

File: core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: let's remove positionIncrementGap too?

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: I think it should be `if (indexSettings.getIndexVersionCreated().onOrAfter(Version.V_6_0_0_alpha3) && settings.get("ignore_case") != null) {` ?

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: This is the wrong place? It should be in the `else` block if `settings.get("tokenizer")` is not null?

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: Please do not load the content of the file in memory. I think it should be easy to do by changing the return type of this method to a `java.io.Reader`?

File: docs/reference/migration/migrate_6_0/mapping.asciidoc
Comment (jpountz) [inline docs/reference/migration/migrate_6_0/mapping.asciidoc]: This does not look true based on the code? We seem to ignore it for new indices all the time (which is fine)?

**Review by jpountz — CHANGES_REQUESTED** [review]
Sorry it is taking so long but I think we are getting closer.

File: core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java
Comment (johtani) [inline core/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzerProvider.java]: Ah, right...

**Comment by johtani** [comment]
@jpountz You don't have to apologize. I learned many things from your comments :) 
I pushed new commit for fixing your comments.

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: let's not log since it is an old index

File: core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/SynonymTokenFilterFactory.java]: I think we should close the analyzer if tokenizerFactory is not null. Otherwise, it is never closed.

```
finally {
    if (tokenizerFactory != null) {
        analyzer.close();
    }
}
```

(this looks like a pre-existing bug, it was not introduced by your PR)

File: docs/reference/migration/migrate_6_0/mapping.asciidoc
Comment (jpountz) [inline docs/reference/migration/migrate_6_0/mapping.asciidoc]: you should have modified `docs/reference/migration/migrate_6_0/mappings.asciidoc` (note the `s` at the end of `mapping`) instead of creating a new file?

**Review by jpountz — COMMENTED** [review]
I left some minor comments. Otherwise it looks good to me!

File: docs/reference/migration/migrate_6_0/mapping.asciidoc
Comment (johtani) [inline docs/reference/migration/migrate_6_0/mapping.asciidoc]: Ah, good catch!

**Review by jpountz — APPROVED** [review]
LGTM

**Pull request merged**
